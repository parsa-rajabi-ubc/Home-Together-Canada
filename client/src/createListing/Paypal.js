/**
 * @Author:     Parsa Rajabi
 * @Created:    2021.2.15
 *
 * @Description: Paypal Component
 *
 */

import React, {useState, useEffect} from 'react';
import {PayPalScriptProvider, PayPalButtons} from "@paypal/react-paypal-js";
import RadioButton from "../common/forms/RadioButton";
import PropTypes from "prop-types";
import Tooltip from "../common/forms/Tooltip";
import {PAYPAL_TEXT} from "../common/constants/TooltipText";


const MESSAGE = {
    THANK_YOU: "Thank you, ",
    PURCHASE: ". Your payment of $",
    ORDER_ID: " was successful! Your OrderID is - ",
    CANCELLED: "Your transaction was cancelled!",
    ERROR: "There was an error with your transaction (you were not charged). Please try again.",
}

export const PAYMENT_STATUS = {
    APPROVED: "Approved",
    CANCELLED: "Cancelled",
    ERROR: "Error",
}

const PAGE_TEXT = {
    TITLE: "Payment",
    LABEL: "Please select how many months you would like to purchase"
};

const purchaseOptionText = [
    {
        // how many months (label)
        label: "1 Month",
        // cost per month (value)
        value: 18
    },
    {
        label: "3 Months",
        value: 45
    },
    {
        label: "6 Months",
        value: 80
    },
    {
        label: "12 Months",
        value: 140
    },
]


const Paypal = (props) => {
    const {paymentStatus, onTransactionOrderIDChange} = props;
    const [numMonthsToPurchase, setNumMonthsToPurchase] = useState();
    const [orderID, setOrderID] = useState();

    useEffect(() => {
        onTransactionOrderIDChange(orderID);
    }, [orderID]);


    function createOrder(data, actions) {
        return actions.order
            .create({
                purchase_units: [
                    {
                        amount: {
                            value: numMonthsToPurchase,
                        },
                    },
                ],
            })
            .then((orderID) => {
                setOrderID(orderID);
                return orderID;
            });
    }

    function onChange(e) {
        setNumMonthsToPurchase(e.target.value);
    }

    const onApprove = (data, actions) => {
        // This function captures the funds from the transaction.
        return actions.order.capture().then(function (details) {
            // This function shows a transaction success message
            alert(MESSAGE.THANK_YOU + details.payer.name.given_name + MESSAGE.PURCHASE + numMonthsToPurchase + MESSAGE.ORDER_ID + details.id);
            paymentStatus(PAYMENT_STATUS.APPROVED);
        });
    }

    const onCancel = () => {
        alert(MESSAGE.CANCELLED);
        paymentStatus(PAYMENT_STATUS.CANCELLED);
    }

    const onError = () => {
        alert(MESSAGE.ERROR);
        paymentStatus(PAYMENT_STATUS.ERROR);
    }

    const paypalOptions = {
        //TODO: Replace this with Twila's Client-id
        "client-id": "AYknWru_AL6Z8b_yEfujCQyL1DAAzhO0aZuQvhfNcqY86bitoAEKQ4iu8Z7bSOjk0BE-tjm8HljYpoba",
        "buyer-country": "CA",
        "currency": "CAD",
        "intent": "capture",
    };

    const purchaseOption = purchaseOptionText.map(
        (option) =>
            <RadioButton key={option.value} label={option.label + " - $" + option.value} name={"months"}
                         value={option.value.toString()} onChange={onChange} blockElement={true}/>
    );
    return (
        <div>
            <h1 className={"page-title mb-5"}> {PAGE_TEXT.TITLE} </h1>
            <p className={"label inline mb-2"}>{PAGE_TEXT.LABEL}</p>
            <Tooltip text={PAYPAL_TEXT.PAYMENT} toolTipID={"paypal"}/>

            {purchaseOption}

            {numMonthsToPurchase &&
            <div className={"w-1/2 m-auto mt-20"}>
                <PayPalScriptProvider options={paypalOptions}>
                    <PayPalButtons style={{layout: "horizontal"}}
                                   createOrder={createOrder}
                                   forceReRender={numMonthsToPurchase}
                                   onApprove={onApprove}
                                   onCancel={onCancel}
                                   onError={onError}
                    />
                </PayPalScriptProvider>
            </div>
            }
        </div>
    )
}

Paypal.propTypes = {
    paymentStatus: PropTypes.func.isRequired,
    onTransactionOrderIDChange: PropTypes.func.isRequired
};

export default Paypal;

