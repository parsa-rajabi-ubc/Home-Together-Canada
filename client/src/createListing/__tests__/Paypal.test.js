/**
 * @Author:     Parsa Rajabi
 * @Created:    2021.2.16
 *
 * @Description: Paypal Test;
 *
 */

import React from 'react';
import renderer from 'react-test-renderer'
import Paypal from "../Paypal";

describe('Paypal', () => {
    describe('Snapshot test', () => {
        it('should match snapshot test regardless of properties', () => {
            // given
            const props = {
                paymentStatus: jest.fn(),
                onTransactionOrderIDChange: jest.fn()
            };
            //when
            const component = renderer.create(<Paypal {...props}/>);
            const tree = component.toJSON();
            //then
            expect(tree).toMatchSnapshot();
        });
    });
});