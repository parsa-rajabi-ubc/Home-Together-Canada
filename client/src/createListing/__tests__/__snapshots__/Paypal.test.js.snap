// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paypal Snapshot test should match snapshot test regardless of properties 1`] = `
<div>
  <h1
    className="page-title mb-5"
  >
     
    Payment
     
  </h1>
  <p
    className="label inline mb-2"
  >
    Please select how many months you would like to purchase
  </p>
  <span>
    <span
      data-for="paypal"
      data-tip={true}
    >
      <svg
        className="inline ml-1 mb-1 text-blue-500 hover:text-blue-400"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2 0a2 2 0 0 0-2 2v12a2 2 0 0 0 2 2h12a2 2 0 0 0 2-2V2a2 2 0 0 0-2-2H2zm3.496 6.033a.237.237 0 0 1-.24-.247C5.35 4.091 6.737 3.5 8.005 3.5c1.396 0 2.672.73 2.672 2.24 0 1.08-.635 1.594-1.244 2.057-.737.559-1.01.768-1.01 1.486v.105a.25.25 0 0 1-.25.25h-.81a.25.25 0 0 1-.25-.246l-.004-.217c-.038-.927.495-1.498 1.168-1.987.59-.444.965-.736.965-1.371 0-.825-.628-1.168-1.314-1.168-.803 0-1.253.478-1.342 1.134-.018.137-.128.25-.266.25h-.825zm2.325 6.443c-.584 0-1.009-.394-1.009-.927 0-.552.425-.94 1.01-.94.609 0 1.028.388 1.028.94 0 .533-.42.927-1.029.927z"
        />
      </svg>
    </span>
    <div
      className="__react_component_tooltip t00000000-0000-0000-0000-000000000000 place-top type-dark"
      data-id="tooltip"
      id="paypal"
    >
      <style
        aria-hidden="true"
        dangerouslySetInnerHTML={
          Object {
            "__html": "
  	.t00000000-0000-0000-0000-000000000000 {
	    color: #fff;
	    background: #222;
	    border: 1px solid transparent;
  	}

  	.t00000000-0000-0000-0000-000000000000.place-top {
        margin-top: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::before {
        border-top: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-top::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        bottom: -6px;
        left: 50%;
        margin-left: -8px;
        border-top-color: #222;
        border-top-style: solid;
        border-top-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-bottom {
        margin-top: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::before {
        border-bottom: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-bottom::after {
        border-left: 8px solid transparent;
        border-right: 8px solid transparent;
        top: -6px;
        left: 50%;
        margin-left: -8px;
        border-bottom-color: #222;
        border-bottom-style: solid;
        border-bottom-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-left {
        margin-left: -10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::before {
        border-left: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-left::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        right: -6px;
        top: 50%;
        margin-top: -4px;
        border-left-color: #222;
        border-left-style: solid;
        border-left-width: 6px;
    }

    .t00000000-0000-0000-0000-000000000000.place-right {
        margin-left: 10px;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::before {
        border-right: 8px solid transparent;
    }
    .t00000000-0000-0000-0000-000000000000.place-right::after {
        border-top: 5px solid transparent;
        border-bottom: 5px solid transparent;
        left: -6px;
        top: 50%;
        margin-top: -4px;
        border-right-color: #222;
        border-right-style: solid;
        border-right-width: 6px;
    }
  ",
          }
        }
      />
      By selecting an option, you're purchasing a certain amount of months that the listing will be live and visible to users on the site
    </div>
  </span>
  <span
    className="block"
  >
    <label
      className="text-base text-gray-700 pr-8"
    >
      <input
        className="form-radio mt-2 ml-2 mr-1 pb-4"
        name="months"
        onChange={[Function]}
        type="radio"
        value="18"
      />
      1 Month - $18
    </label>
  </span>
  <span
    className="block"
  >
    <label
      className="text-base text-gray-700 pr-8"
    >
      <input
        className="form-radio mt-2 ml-2 mr-1 pb-4"
        name="months"
        onChange={[Function]}
        type="radio"
        value="45"
      />
      3 Months - $45
    </label>
  </span>
  <span
    className="block"
  >
    <label
      className="text-base text-gray-700 pr-8"
    >
      <input
        className="form-radio mt-2 ml-2 mr-1 pb-4"
        name="months"
        onChange={[Function]}
        type="radio"
        value="80"
      />
      6 Months - $80
    </label>
  </span>
  <span
    className="block"
  >
    <label
      className="text-base text-gray-700 pr-8"
    >
      <input
        className="form-radio mt-2 ml-2 mr-1 pb-4"
        name="months"
        onChange={[Function]}
        type="radio"
        value="140"
      />
      12 Months - $140
    </label>
  </span>
</div>
`;
